src/SignalR/common/Shared/AsyncEnumerableAdapters.cs

```diff
@@ -66,7 +67,16 @@ public IAsyncEnumerator<TResult> GetAsyncEnumerator(CancellationToken cancellati
                 {
                     var registration = cancellationToken.Register((ctsState) =>
                     {
-                        ((CancellationTokenSource)ctsState).Cancel();
+                        try
+                        {
+                            ((CancellationTokenSource)ctsState).Cancel();
+                        }
+                        catch (ObjectDisposedException)
+                        {
+                            // cancellationToken is passed in by the user, and if they call cancel after the
+                            // enumerator is finished but before the enumerator is disposed
+                            // then _cts might already be disposed by our wrapping code
```

Comment:
I don't think we should be throwing first-chance exceptions for a fairly common usage pattern. Can we combine MakeCancelableTypedAsyncEnumerable with CastIAsyncEnumerable in a way that ensures the registration is always disposed before CastIAsyncEnumerable disposes the CTS? The current layering makes it difficult, but if those were combined it should be easy enough.

We could also consider not disposing the CTS at all, but I'd prefer disposing it if it's straightforward enough.

Reply:
I went with the second option as fixing the layering will probably lead to a lot of duplicate code. And layering asyncenumerables on top of each other started annoying me 😆